### DISCUSSION TITLE
What all is this supposed to pick up from the Tigo unit?

### DISCUSSION THREAD

**Discussion by zestysoft** [mletenay/home-assistant-tigo#4] [Q&A]
Under "Configuration", I see the ability to toggle the modules on or off, and under "Sensors", I see "CCA cca_temperature", but I don't see anything else like panel details.  I know this is a new project, so maybe this is still WIP?  Otherwise, I'm going to guess the different firmware versions are breaking things?

**Comment by zestysoft** [comment]
Looks like this project isn't fully compatible with the 2.8.5 firmware.  There are lots of examples in the code that call non-existing cgi-bin endpoints.

FWIW, these are all of the cgi-bin options with 2.8.5:

`
Fronius
adt
algorithm
answers
autoconfig
autorandevu
autorepeater
basememory
beacons
calmodefixer.cgi
channels
cleanup
cmdline
counters
ctrldatasync
daqsweeper.cgi
datetime
ddr
discovery
discresults
discui
enemies
fixes
friends
fronius_power
gatewaypower.cgi
gateways
gatewaytuner
gpioweb
gtwaytest
gwdiags
gwstat
info
inisettings
inverterconfig
inverterstatus
ivcurve
ivcurvechart
ivcurveui
lastknowngood
lmuchart
lmudata
lmudui
lmus
lmustatus3
meterconfig
mm2esfixer.cgi
mmdstatus
mmepstatus
mobile_api
multimmu
network
odt
opentestfixtureport
pan
pleasemesh
pleasereboot
pleasetcp
rawdata
remup.cgi
remupui
repeaterconfig
repeaters
reslotter
routing
sendbeacon
sendmessage
shell
smrt
softrestart.cgi
stat
summary
summary_config
summary_data
summary_energy
summary_energy.cron
summary_jsconfig
syncfet
sysconfig
tcp2serial
tcp2serialconsole
tester
timezone
virtvars
vout
wifi
wlconfig
`

**Reply by mletenay** [reply]
Hmm, I wrote and test it against 3.7.6 version and at this moment it uses lmudui, meshdatapower, meshnodeinfo, meshnodever pages.
But if you save me the html of your pages, I should be able to write 2.8.x compatible parsers.

**Reply by mletenay** [reply]
Don't you have any screen which resembles this one ?
![image](https://github.com/mletenay/home-assistant-tigo/assets/4719495/2f4b3004-e75f-4ef0-88fd-01c6f7250018)
or this 
![image](https://github.com/mletenay/home-assistant-tigo/assets/4719495/534dc8a4-1dcd-4872-9937-84c8b60e8637)

**Reply by zestysoft** [reply]
This is /cgi-bin/mmdstatus

<img width="1548" alt="Screenshot 2024-07-01 at 3 01 00 PM" src="https://github.com/mletenay/home-assistant-tigo/assets/742554/48e5b1b1-c168-4467-a817-8c234d19aab8">
and this is /cgi-bin/stat
<img width="1541" alt="Screenshot 2024-07-01 at 3 04 45 PM" src="https://github.com/mletenay/home-assistant-tigo/assets/742554/c36b3599-5104-4c5b-8eae-1fedeb31abab">
and this is /cgi-bin/rawdata
<img width="1493" alt="Screenshot 2024-07-01 at 3 32 36 PM" src="https://github.com/mletenay/home-assistant-tigo/assets/742554/ff27937c-d9a3-48bb-84f9-8524fc617304">


I can send ya the raw html from any of these if they work out for ya.

**Reply by mletenay** [reply]
The mmdstatus is feasible workaround. I actually started with that on 3.x., but then I switched to meshdatapower which seems to be bit more reliable. But it will do if you send me the raw html from the 2.x fw.
As for the rest, they are optional - I just parse version and mac for the diagnostic purposes, so strictly not necessary, just nice to have.

**Reply by zestysoft** [reply]
I should probably wait until it's sunny and send that instead, but I'll send this in the meantime.  Thanks for looking into this!
[mmdstatus.zip](https://github.com/user-attachments/files/16095267/mmdstatus.zip)

**Reply by mletenay** [reply]
OK, I made the necessary changes and it should work with 2.8.x firmwares now.
Let me know if there's any problem.

**Reply by zestysoft** [reply]
I've updated the HACS plugin but it looks like there's some trouble:

```
2024-07-06 23:32:06.921 ERROR (MainThread) [homeassistant.config_entries] Error setting up entry Tigo CCA for tigo
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 594, in async_setup
    result = await component.async_setup_entry(hass, self)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/tigo/__init__.py", line 33, in async_setup_entry
    coordinator = TigoUpdateCoordinator(hass, entry, cca, status)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/tigo/coordinator.py", line 40, in __init__
    p.label: _panel_device(status, p) for p in cca.panels.values()
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/tigo/coordinator.py", line 72, in _panel_device
    model=panel.model(),
          ^^^^^^^^^^^^^
  File "/config/custom_components/tigo/tigo_cca.py", line 21, in model
    if "455" in self.hw:
       ^^^^^^^^^^^^^^^^
TypeError: argument of type 'NoneType' is not iterable
```
basic_test.py returns this:
```
2024-07-06 16:38:58,714 __init__(64) - DEBUG: Using selector: KqueueSelector
{'A1': PanelVersionInfo(label='A1', mac='04C05B0010404633', fw=None, hw=None), 'A2': PanelVersionInfo(label='A2', mac='04C05B00104049D6', fw=None, hw=None), 'A3': PanelVersionInfo(label='A3', mac='04C05B00104049BC', fw=None, hw=None), 'A4': PanelVersionInfo(label='A4', mac='04C05B0010404A78', fw=None, hw=None), 'A5': PanelVersionInfo(label='A5', mac='04C05B00104049A8', fw=None, hw=None), 'A6': PanelVersionInfo(label='A6', mac='04C05B0010404A59', fw=None, hw=None), 'A7': PanelVersionInfo(label='A7', mac='04C05B00104049BE', fw=None, hw=None), 'A8': PanelVersionInfo(label='A8', mac='04C05B00104049E2', fw=None, hw=None), 'A9': PanelVersionInfo(label='A9', mac='04C05B001040283A', fw=None, hw=None), 'A10': PanelVersionInfo(label='A10', mac='04C05B001040413F', fw=None, hw=None), 'A11': PanelVersionInfo(label='A11', mac='04C05B00104046EF', fw=None, hw=None), 'A12': PanelVersionInfo(label='A12', mac='04C05B0010404A05', fw=None, hw=None), 'B1': PanelVersionInfo(label='B1', mac='04C05B001040457E', fw=None, hw=None), 'B2': PanelVersionInfo(label='B2', mac='04C05B00104046DA', fw=None, hw=None), 'B3': PanelVersionInfo(label='B3', mac='04C05B00104046E3', fw=None, hw=None), 'B4': PanelVersionInfo(label='B4', mac='04C05B00104035E7', fw=None, hw=None), 'B5': PanelVersionInfo(label='B5', mac='04C05B00104049EA', fw=None, hw=None), 'B6': PanelVersionInfo(label='B6', mac='04C05B0010402826', fw=None, hw=None), 'B7': PanelVersionInfo(label='B7', mac='04C05B0010404A5A', fw=None, hw=None), 'B8': PanelVersionInfo(label='B8', mac='04C05B0010404613', fw=None, hw=None)}
Tigo CCA #04C05B805BFC
FW: 2.8.5-nd-bin
HW: Tigo Original Cloud Connect
Temp: 37.3

Panel A2: Pwr=101.79, Tmp=40.2
Panel A3: Pwr=104.59, Tmp=39.7
Panel A4: Pwr=106.06, Tmp=40.5
Panel A5: Pwr=107.12, Tmp=None
Panel A8: Pwr=104.07, Tmp=38.5
Panel A9: Pwr=104.68, Tmp=39.1
Panel A10: Pwr=104.43, Tmp=38.2
Panel A11: Pwr=104.07, Tmp=38.4
Panel A12: Pwr=106.54, Tmp=35.5
Panel B5: Pwr=118.29, Tmp=39.5
Panel B7: Pwr=105.9, Tmp=36.0
Panel B8: Pwr=105.31, Tmp=None
```

**Reply by mletenay** [reply]
Ooops. Fixed. Try now.

**Reply by zestysoft** [reply]
Much better!  I'm seeing all the correct data from the cca and the panels, however the devices shown in HA isn't a complete picture -- only 16 devices instead of 21 (I've got 20 panels, and the CCA).

I know that the communication of these panels are hit-n-miss when I look at the data from cgi-bin/mmdstatus -- like I have to hit refresh a number of times before I can see the data from the panel I'm interested in -- is there a chance that when this initialized, it just grabbed the ones that happened to return data?

**Reply by mletenay** [reply]
Yeah, it behaves indeed like this. Upon start it collects panel data and then just refreshes them.
Mine system behaves reliably, all panels report all the time. It's rather weird 2.8.x does not.
Isn't there any other page which lists panels and is more reliable ?

**Reply by zestysoft** [reply]
Not that I could find.  There are other tables that do contains all of the panel details though, so maybe that could be used to initialize instead?
for instance cgi-bin/lmus provides this data:
[lmus.zip](https://github.com/user-attachments/files/16122627/lmus.zip)

**Reply by zestysoft** [reply]
I should mention that I've restarted HA a number of times since this, and it has picked up most of the other panels.  I'm fine redoing this until all of the panels are seen instead of whatever additional work this might put on you for who knows how many people still use this older equipment.

Do you know how often it polls the Tigo CCA to get the energy data? Given how unreliable that endpoint is, I figure maybe I should make that more aggressive?

**Reply by zestysoft** [reply]
I stand corrected -- when I checked the data today (11:37am local time) it was only picking up data from a single panel. I had hoped there was some type of persistence that happened as panels were discovered, but apparently not.

**Reply by mletenay** [reply]
Is also the /lmus so unreliable or just /mmdstatus ?
(My lmus, which is not reachable from menu but exists, displays date Sat Oct 14 21:13:49 2023. Which is most likely date when I installed the tigo stuff, so it smells like pretty static/reliable).

**Reply by zestysoft** [reply]
Oh good catch with the time stamp -- yeah, mine shows when it was installed too, so I assume it's reliable. I haven't seen an instance where any of that info has gone missing when viewed.

**Reply by zestysoft** [reply]
Just wanted to follow up, the missing fields are not a web issue, that's how the data is stored in the daq files.  Since Tigo uses something akin to zigbee to wireless transmit the data, I assume some setups like mine may have more trouble than others gathering data if there's wireless interference.  I'm curious what kind of signal strength your setup has and if newer tigo firmware averages out or maintains the last known value instead.
